### PR TITLE
feature: implement a new `TableBuilder` field

### PR DESCRIPTION
**Pull request by ryangjchandler**
Started working on this on a stream, see screenshot below:
CleanShot 2022-06-04 at 21 18 03@2x

Still a few things left to do:
* [x] Check dark mode support
* [x] Use translations for button labels
* [x] Allow changing button labels
* [ ] Run design past someone with a better eye than me
* [x] Move Alpine code into JS file
* [x] a11y

### CODE REVIEW

**Comment by themahdavi** [comment]
Hey @ryangjchandler about remove icons some issue 

I use `heroicon-o-trash` but on repeater and key-value component use `heroicon-s-trash` 
Is good if fix this small ui issue

And other thing good enough to finished up

And some idea : can use another field type on each cells

**Comment by ryangjchandler** [comment]
@themahdavi I think it should be fixed to text inputs, the same as `KeyValue`.

I'll make a note to change the icons around.

**Comment by JackWH** [comment]
@ryangjchandler This looks like a fantastic addition to Filament 🚀 Any updates on this PR? I'd love to use this in a project I'm working on at the moment — if you need any help finishing it off, let me know :)

**Comment by themahdavi** [comment]
Hey @ryangjchandler when tonight arrive ? 😅😅

**Comment by ryangjchandler** [comment]
Sorry. I've had some other priorities.

File: packages/forms/resources/js/components/table-builder.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            this.state = this.state.filter((_, i) => i !== row)
+        },
+
+        removeColumn(column) {
+            this.columns--
+
+            this.state = this.state.map((row) => {
+                return row.filter((_, i) => i !== column - 1)
+            })
+        },
```
Comment (zepfietje) [inline packages/forms/resources/js/components/table-builder.js:43]:
```suggestion
        removeColumn(columnToRemove) {
            this.columns--

            this.state = this.state.map((row) => {
                return row.filter((_, column) => column !== columnToRemove - 1)
            })
        },
```

File: packages/forms/resources/js/components/table-builder.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                    }
+
+                    this.state[i][column] = null
+                }
+            }
+        },
+
+        removeRow(row) {
+            this.state = this.state.filter((_, i) => i !== row)
+        },
```
Comment (zepfietje) [inline packages/forms/resources/js/components/table-builder.js:35]:
```suggestion
        removeRow(rowToRemove) {
            this.state = this.state.filter((_, row) => row !== rowToRemove)
        },
```

File: packages/forms/resources/lang/en/components.php [hunk 3]
```diff
@@ -214,6 +214,22 @@
 
     ],
 
+    'table_builder' => [
+
+        'buttons' => [
+
+            'add_row' => [
+                'label' => 'Add Row',
```
Comment (zepfietje) [inline packages/forms/resources/lang/en/components.php:222]:
```suggestion
                'label' => 'Add row',
```

File: packages/forms/resources/lang/en/components.php [hunk 4]
```diff
@@ -214,6 +214,22 @@
...
+            ],
+
+            'add_column' => [
+                'label' => 'Add Column',
```
Comment (zepfietje) [inline packages/forms/resources/lang/en/components.php:226]:
```suggestion
                'label' => 'Add column',
```

File: packages/forms/resources/views/components/table-builder.blade.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+                            @if($deleteRows && ! $isDisabled())
+                                <td x-show="state.length > 1" class="p-0">
+                                    <div class="flex items-center justify-center">
+                                        <button
+                                            x-on:click="removeRow(index)"
+                                            type="button"
+                                            class="text-gray-400 hover:text-gray-600 focus:text-gray-600"
+                                        >
+                                            <x-heroicon-o-x class="w-4 h-4" />
```
Comment (zepfietje) [inline packages/forms/resources/views/components/table-builder.blade.php:96]: What if we use this?
```suggestion
                                            <x-heroicon-s-x class="w-5 h-5" />
```

File: packages/forms/resources/views/components/table-builder.blade.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                                        </button>
+                                    @endif
+                                </div>
+                            </th>
+                        </template>
+
+                        @if($deleteRows && ! $isDisabled())
+                            <th x-show="state.length > 1" class="w-10">
+
```
Comment (zepfietje) [inline packages/forms/resources/views/components/table-builder.blade.php:60]: Doe we need this blank line?

File: packages/forms/src/Components/TableBuilder.php [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    protected string | Closure $addColumnButtonLabel;
+
+    protected string | Closure $addRowButtonLabel;
+
+    protected int | Closure $defaultRows = 1;
+
+    protected int | Closure $defaultColumns = 1;
+
+    protected bool | Closure $shouldDisableAddingRows = false;
```
Comment (zepfietje) [inline packages/forms/src/Components/TableBuilder.php:23]: What about renaming these to be consistent with the rest of the codebase?
```suggestion
    protected bool | Closure $isAddingRowsDisabled = false;
```

**Review by zepfietje — CHANGES_REQUESTED** [review]

File: packages/forms/src/Components/TableBuilder.php (hunk 7 above)
Comment (danharrin) [reply packages/forms/src/Components/TableBuilder.php:23]: $canAddRows ?

**Comment by saschagysel** [comment]
Hey Ryang, will you find time to finish this? Its a feature I would really like to use in my product. Thanks a lot anyways.

**Comment by ziming** [comment]
Hi, is there anything we can do to help get this merged?

**Review by Diddyy — COMMENTED** [review]
I would be willing to help if needed! :)

File: packages/forms/resources/views/components/table-builder.blade.php [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                            </th>
```
Comment (zepfietje) [inline packages/forms/resources/views/components/table-builder.blade.php]:
```suggestion
                            </th>
```

**Review by zepfietje — CHANGES_REQUESTED** [review]

**Comment by zepfietje** [comment]
@danharrin @ryangjchandler, what do we want to do with this PR?

**Comment by danharrin** [comment]
I'm not sure when Ryan is going to finish this up, and I'm erring on the side of suggesting it should be a plugin anyway. If he wants / finds time to finish, we can have that discussion of whether or not it belongs in core.

**Comment by danharrin** [comment]
I'll delete the branch, but if this ends up being worked on then you should be able to just `Restore branch`

**Comment by zepfietje** [comment]
Agree, @danharrin.

**Pull request closed without merging**